### PR TITLE
Add a forward authentication option

### PR DESCRIPTION
**Pull request by diegooliveira**
Now it is possible to delegate the authentication to a third party agent, like a OAuth endpoint. This authentication option extracts parameters from the original request, forward then to a authentication endpoint, if it succeed part of the response body might be added to the request.

### CODE REVIEW

**Comment by emilevauge** [comment]
Ping @diegooliveira, tests are failing on Travis ;) Could you also rebase your PR?

**Comment by WTFKr0** [comment]
Hi,

Do you think this could be done with a JWT auth.
Like the docker registry works
Conf example of a docker registry delagate auth to a gitlab instance : 
```
cat /root/docker/volumes/registry/config/config.yml
version: 0.1
log:
  fields:
    service: registry
storage:
    cache:
        blobdescriptor: inmemory
    filesystem:
        rootdirectory: /var/lib/registry
    delete:
        enabled: true
http:
    addr: :5000
    headers:
        X-Content-Type-Options: [nosniff]
health:
  storagedriver:
    enabled: true
    interval: 10s
    threshold: 3
auth:
  token:
    realm: https://gitlab.mydomain.org/jwt/auth
    service: container_registry
    issuer: omnibus-gitlab-issuer
    rootcertbundle: /certs/domain.crt
```

Thanks for reading

**Comment by diegooliveira** [comment]
@WTFKr0 I'm reading a little more about the jwt authentication schema to get the use case. As my understand to date you need a way to have the secret stored in some sort of trusted store, this way you are able to validate if the Authentication header is valid.

I don't know if you are planning to add specific knowledge about the authentication schema in traefic, if so implementing this functionality makes a lot of sense. The current implementation just **forwards** the desired tokens(headers, parameters, etc) to a remote server and all the logic happens there, this might give a little more flexibility allowing any implementation to be pluggable.

Going in a auth schema specific for "jwt" it would be nice to have a configuration like "entryPoints.http.auth.jwt", but this implementation don't exclude the forward one. If you wish I may do a new PR implementing the jwt logic.

**Comment by emilevauge** [comment]
@diegooliveira +1 on focusing on authentication forwarding only for now. I'm not even sure that it would be a good thing to do more in Traefik ^^

**Comment by diegooliveira** [comment]
@emilevauge @WTFKr0 rebase over master and test fixed

**Comment by emilevauge** [comment]
@diegooliveira your last rebase is not that great, you added a lot of merge files `glide_*` ;)

**Comment by diegooliveira** [comment]
@emilevauge cleaned up

**Comment by meggarr** [comment]
Hi @diegooliveira , when traefik is deployed in Kubernetes, can the *Forward* authentication be configured via IngressController's annotation?

**Comment by diegooliveira** [comment]
@meggarr I don't know that much of Kubernets, by what i did read, the ingress controller is a way to have a load balance in your deployment and traefik is a possible solution for this setup (https://docs.traefik.io/user-guide/kubernetes/). If you are running this configuration I think the only thing you need is a template that writes the traefic forward authentication configuration.

**Comment by meggarr** [comment]
@diegooliveira yes, I have successfully deployed *Traefik* IngressController in my k8s cluster, the backends (k8s services) are automatically collected by *Ingress* info. You know it has an annotation `traefik.frontend.rule.type: pathprefixstrip` to do a path based routing.

I'm just thinking with *Forward* auth feature, probably it can provide new annotations for `frontend.http.auth.forward.xxxx` to forward the auth to a k8s service, with a service name and path.

Maybe a single annotation is not enough, or there's a better way, I am just asking to have a discussion..

Thanks!

**Comment by emilevauge** [comment]
@diegofernandes sorry, could you rebase again?

File: auth/forward.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	"github.com/containous/traefik/types"
+	"github.com/stretchr/stew/objects"
+)
+
+// Forward the authentication to a external server
+func Forward(forward *types.Forward, w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+
+	client := http.Client{}
+
+	forwardReq, _ := http.NewRequest("GET", forward.Address, nil)
```
Comment (emilevauge) [inline auth/forward.go]: shadowing err here is not a good idea ;)

Comment (diegooliveira) [inline auth/forward.go]: @emilevauge any idea of a good error handling for this case, is it OK to return an 5x error code?

Comment (emilevauge) [inline auth/forward.go]: @diegooliveira I think it is. WDYT @containous/traefik ?

File: auth/forward.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	forwardQuery := forwardReq.URL.Query()
+	rQuery := r.URL.Query()
+	for _, reqParam := range forward.RequestParameters {
+		paramValue := rQuery.Get(reqParam.Name)
+		forwardQuery.Add(reqParam.As, paramValue)
+	}
+	forwardReq.URL.RawQuery = forwardQuery.Encode()
+	forwardResponse, forwardErr := client.Do(forwardReq)
+	if forwardErr != nil || forwardResponse.StatusCode != 200 {
+		log.Debugf("Auth failed...")
```
Comment (emilevauge) [inline auth/forward.go]: I think forwardErr != nil is not the same as forwardResponse.StatusCode != 200.
Also, you should sent back an error to the client, right?

Comment (diegooliveira) [inline auth/forward.go]: @emilevauge now it's handled different from each other.

**Comment by Toflar** [comment]
I would like to join the discussion because I'm looking for a way to integrate a custom auth back end into traefik and this PR looks like it's half of what I need :-) I feel like it's a bit too much focussing on "e-mail" and "token" because in my opinion authorization can be achieved by virtually implementation. Whether it is a combination of e-mail and password or token, http headers, a custom nonsense legacy implementation or a standardized way like JWT.

Wdyt about implementing something like the nginx [ngx_http_auth_request_module](http://nginx.org/en/docs/http/ngx_http_auth_request_module.html)? Basically this just forwards the HTTP request and based on the response status code then either allows or denies access. No extraction of specific fields or other magic. This would allow for arbitrary auth back ends.
As a bonus we should maybe use a library that supports caching of the auth responses when forwarding the requests so that it is possible for back end auth implementations to send regular HTTP caching headers and as such prevent traefik from forwarding the request every single time even if a token could be e.g. cached for 15 minutes.

Unfortunately, I'm not coding in Go (yet) so I can just share my thoughts on this one :-)

**Comment by emilevauge** [comment]
I agree with @Toflar on this, I think it would be great to mimic the nginx ngx_http_auth_request_module.
WDYT @diegooliveira ?
Another comment: not sure about your `{}` in your `log.Debugf()` logs.

**Comment by diegooliveira** [comment]
@toflar @emilevauge I'm in vacation this week, comming back to the world next Monday. I have a point in forward everything to a auth server. The payload might contain sensitive information (it's my case). I'll take a look in the nginx module. What do you think of a confirmation to skip sending the body?

**Comment by Toflar** [comment]
Hey @diegooliveira, no problem, enjoy your time :-)

> What do you think of a configuration to skip sending the body?

Well, an auth back end by definition always deals with sensitive information, doesn't it? :-) So it should be secured by default. I mean we could even deny auth back ends that do not send via `https` :-)
But still I think yes, a configuration sounds valid to me. It would reduce bandwidth usage if you don't need the body for authentication and headers are enough (which is very often the case: basic auth, api tokens etc.). So yeah, excellent input! 👍

**Comment by burdandrei** [comment]
@diegooliveira auth option is really good, but shouldn't there be an option to disable for specific frontend?

**Comment by vladshub** [comment]
Hi @Toflar,
It would be nice to be able to pass headers and content from the body but it could (even should) happen down the road as another PR...

**Comment by emilevauge** [comment]
Ping @diegooliveira :)

**Comment by emilevauge** [comment]
@diegooliveira A new PR has been submitted carrying this one https://github.com/containous/traefik/pull/1030.

**Pull request closed without merging**